tests/fuzzers/bls12381/bls12381_fuzz.go

```diff
@@ -0,0 +1,250 @@
+	}
+
+	return 1
+}
+
+func FuzzCrossG1MultiExp(data []byte) int {
+	// result = multiExp(bases, scalars)
+	// n random scalars
+
+	// we take the first byte to derive n (max = 17)
+	if len(data) < 1 {
+		return 0
+	}
+	n := (uint8(data[0]) % 16) + 1
+
+	input := bytes.NewReader(data[1:])
+
+	// n random scalars
+	scalars := make([]*big.Int, n)
+	_scalars := make([]fr.Element, n)
+	for i := 0; i < int(n); i++ {
+		// note that geth/crypto/bls12381 works only with scalars <= 32bytes
+		s, err := randomScalar(input, fr.Modulus())
+		if err != nil {
+			return 0
+		}
+		scalars[i] = s
+		_scalars[i].SetBigInt(scalars[i]).FromMont()
+	}
+
+	// get a random G1 point as basis
+	kp1, cp1, err := getG1Points(input)
+	if err != nil {
+		return 0
+	}
+	kps := make([]*bls12381.PointG1, n)
+	cps := make([]gnark.G1Affine, n)
+	for i := 0; i < int(n); i++ {
+		kps[i] = new(bls12381.PointG1).Set(kp1)
+		cps[i].Set(cp1)
+	}
+
+	// compute multi exponentiation
+	g1 := bls12381.NewG1()
+	kp := bls12381.PointG1{}
+	if _, err := g1.MultiExp(&kp, kps, scalars); err != nil {
+		panic(fmt.Sprintf("G1 multi exponentiation errored (geth): %v", err))
+	}
+	// note that geth/crypto/bls12381.MultiExp mutates the scalars slice (and sets all the scalars to zero)
+
+	// gnark multi exp
+	cp := new(gnark.G1Affine)
+	cp.MultiExp(cps, _scalars)
+
+	// compare result
+	if !(bytes.Equal(cp.Marshal(), g1.ToBytes(&kp))) {
+		panic("G1 multi exponentiation mismatch gnark / geth ")
+	}
+
+	return 1
+}
+
+func getG1Points(input io.Reader) (*bls12381.PointG1, *gnark.G1Affine, error) {
+	// sample a random scalar
+	s, err := randomScalar(input, fp.Modulus())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// compute a random point
+	cp := new(gnark.G1Affine)
+	_, _, g1Gen, _ := gnark.Generators()
+	cp.ScalarMultiplication(&g1Gen, s)
+	cpBytes := cp.Marshal()
+
+	// marshal gnark point -> geth point
+	g1 := bls12381.NewG1()
+	kp, err := g1.FromBytes(cpBytes)
+	if err != nil {
+		panic(fmt.Sprintf("Could not marshal gnark.G1 -> geth.G1: %v", err))
+	}
+	if !bytes.Equal(g1.ToBytes(kp), cpBytes) {
+		panic("bytes(gnark.G1) != bytes(geth.G1)")
+	}
+
+	return kp, cp, nil
+}
+
+func getG2Points(input io.Reader) (*bls12381.PointG2, *gnark.G2Affine, error) {
+	// sample a random scalar
+	s, err := randomScalar(input, fp.Modulus())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// compute a random point
+	cp := new(gnark.G2Affine)
+	_, _, _, g2Gen := gnark.Generators()
+	cp.ScalarMultiplication(&g2Gen, s)
+	cpBytes := cp.Marshal()
+
+	// marshal gnark point -> geth point
+	g2 := bls12381.NewG2()
+	kp, err := g2.FromBytes(cpBytes)
+	if err != nil {
+		panic(fmt.Sprintf("Could not marshal gnark.G2 -> geth.G2: %v", err))
+	}
+	if !bytes.Equal(g2.ToBytes(kp), cpBytes) {
+		panic("bytes(gnark.G2) != bytes(geth.G2)")
+	}
+
+	return kp, cp, nil
+}
+
+func randomScalar(r io.Reader, max *big.Int) (k *big.Int, err error) {
+	for {
+		k, err = rand.Int(r, max)
```

Comment:
using `rand` here is no good. It makes the fuzzer execution non-deterministic based on the input. So if the fuzzer does find a fault, then the input testcase won't produce the same path if we try to reproduce it. It can be solved in two ways:
1. Use a prng seeded from the input, or 
2. Read the `k` directly from the reader. 

The second option should be preferred -- unless we need to read huge amounts of data

Reply:
Sorry, nevermind, `rand.int(io.Reader., ..)` does the right thing. I confused it with another method